grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/atom/binary/RelationshipAtom.java

```diff
@@ -761,73 +726,68 @@ private RelationshipAtom inferRoles(Answer sub){
-                                        SchemaConcept parentType = parentVarSchemaConceptMap.get(e.getKey().getRolePlayer().var());
-                                        SchemaConcept childType = childVarSchemaConceptMap.get(e.getValue().getRolePlayer().var());
-                                        return !(parentType != null && parentType.equals(childType));
-                                    }))
-                                    //prioritise mappings with sam var substitution (idpredicates)
-                                    .sorted(Comparator.comparing(e -> {
-                                        IdPredicate parentId = parentAtom.getPredicates(IdPredicate.class)
-                                                .filter(p -> p.getVarName().equals(e.getKey().getRolePlayer().var()))
-                                                .findFirst().orElse(null);
-                                        IdPredicate childId = getPredicates(IdPredicate.class)
-                                                .filter(p -> p.getVarName().equals(e.getValue().getRolePlayer().var()))
-                                                .findFirst().orElse(null);
-                                        return !(parentId != null && parentId.isEquivalent(childId));
-                                    }))
-                                    .findFirst().orElse(null)
-                    );
-
-            RelationPlayer parentCasting = rpPair.getKey();
-            RelationPlayer childCasting = rpPair.getValue();
-            rolePlayerMappings.add(new Pair<>(childCasting, parentCasting));
-
-            //remove corresponding entries
-            Pair<RelationPlayer, List<RelationPlayer>> entryToRemove = compatibleMappings.stream()
-                    .filter(e -> e.getKey() == parentCasting)
-                    .findFirst().orElse(null);
-            compatibleMappings.remove(entryToRemove);
-            compatibleMappings.stream()
-                    .filter(e -> e.getValue().contains(childCasting))
-                    .forEach(e -> e.getValue().remove(childCasting));
-        }
-        return rolePlayerMappings;
+        return Sets.cartesianProduct(compatibleMappingsPerParentRP).stream()
+                .filter(list -> !list.isEmpty())
+                //check the same child rp is not mapped to multiple parent rps
+                .filter(list -> {
+                    List<RelationPlayer> listChildRps = list.stream().map(Pair::getKey).collect(Collectors.toList());
+                    //NB: this preserves cardinality instead of removing all occuring instances which is what we want
+                    return ReasonerUtils.subtract(listChildRps, this.getRelationPlayers()).isEmpty();
+                })
+                //check all parent rps mapped
+                .filter(list -> list.stream().map(Pair::getValue).collect(Collectors.toList()).containsAll(parentAtom.getRelationPlayers()))
```

Comment:
Can you break this down better? This filter step is very hard to follow. Even just creating a method for it would help.